File: frontend/src/lib/notifications.ts
import { apiFetch } from "./api";

export interface NotificationItem {
  id: number;
  type: string;
  severity: "info" | "warning" | "error" | "critical";
  title: string;
  message: string;
  entity_ref: string | null;
  read_at: string | null;
  created_at: string;
}

export async function fetchNotifications(unreadOnly = false, limit = 50): Promise<NotificationItem[]> {
  const params = new URLSearchParams();
  if (unreadOnly) params.set("unread_only", "true");
  params.set("limit", String(limit));
  return apiFetch<NotificationItem[]>(`/api/v1/notifications?${params}`);
}

export async function fetchUnreadCount(): Promise<number> {
  const r = await apiFetch<{ count: number }>("/api/v1/notifications/unread-count");
  return r.count;
}

export async function markRead(id: number): Promise<void> {
  await apiFetch(`/api/v1/notifications/${id}/read`, { method: "PATCH" });
}

export async function markAllRead(): Promise<void> {
  await apiFetch("/api/v1/notifications/read-all", { method: "PATCH" });
}

export async function deleteNotification(id: number): Promise<void> {
  await apiFetch(`/api/v1/notifications/${id}`, { method: "DELETE" });
}

export function severityColor(severity: string): string {
  switch (severity) {
    case "critical": return "var(--status-bad)";
    case "error":    return "var(--status-bad)";
    case "warning":  return "var(--status-warn)";
    default:         return "var(--status-info)";
  }
}

File: app/api/v1/notifications.py
"""Notifications API — list, count unread, mark read."""
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy import func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.api.deps_auth import get_current_user
from app.db.models.notifications import Notification
from app.db.models.user import User

router = APIRouter(prefix="/notifications")


class NotificationOut(BaseModel):
    id: int
    type: str
    severity: str
    title: str
    message: str
    entity_ref: Optional[str]
    read_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


@router.get("", response_model=list[NotificationOut])
async def list_notifications(
    unread_only: bool = Query(False),
    limit: int = Query(50, le=200),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    stmt = select(Notification).where(Notification.user_id == current_user.id)
    if unread_only:
        stmt = stmt.where(Notification.read_at.is_(None))
    stmt = stmt.order_by(Notification.created_at.desc()).limit(limit)
    res = await db.execute(stmt)
    return list(res.scalars().all())


@router.get("/unread-count")
async def unread_count(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    res = await db.execute(
        select(func.count())
        .select_from(Notification)
        .where(
            Notification.user_id == current_user.id,
            Notification.read_at.is_(None),
        )
    )
    return {"count": res.scalar_one()}


@router.patch("/{notification_id}/read")
async def mark_read(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await db.execute(
        update(Notification)
        .where(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
        .values(read_at=datetime.now(timezone.utc))
    )
    await db.commit()
    return {"ok": True}


@router.patch("/read-all")
async def mark_all_read(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await db.execute(
        update(Notification)
        .where(
            Notification.user_id == current_user.id,
            Notification.read_at.is_(None),
        )
        .values(read_at=datetime.now(timezone.utc))
    )
    await db.commit()
    return {"ok": True}


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    res = await db.execute(
        select(Notification).where(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
    )
    notif = res.scalar_one_or_none()
    if notif:
        await db.delete(notif)
        await db.commit()
    return {"ok": True}
